File: apps/backend/src/wise_workbench/api/routers/knowledge.py
"""Guidance and the knowledge hub (RK-2, RK-3, RK-4)."""

from __future__ import annotations

from typing import Annotated, Literal

from fastapi import APIRouter, Query

from wise_workbench.api import schemas
from wise_workbench.api.deps import ContainerDep

router = APIRouter(prefix="/projects/{projectId}", tags=["knowledge"])

Kind = Literal["layer", "constraint", "expectation", "failure_mode"]


@router.get(
    "/knowledge/hub",
    operation_id="getKnowledgeHub",
    response_model=schemas.HubIndex,
    responses={404: {"model": schemas.Problem}},
    description=(
        "The knowledge hub of the project's process pack as node and edge tables: stages, expectation areas, "
        "expectations, failure modes, usual reasons, usual actions and KPIs, each with its plain name and its "
        "method name. Nodes the project has written its own note on are marked `hasOverlay`."
    ),
)
def get_knowledge_hub(
    projectId: str,
    c: ContainerDep,
    process: Annotated[str | None, Query(description="pack id; default: the project's process")] = None,
) -> schemas.HubIndex:
    return schemas.HubIndex(**c.knowledge.hub(projectId, process))


@router.get(
    "/knowledge/hub/{nodeId:path}",
    operation_id="getKnowledgeHubPage",
    response_model=schemas.HubPage,
    responses={404: {"model": schemas.Problem}},
    description=(
        "One hub page: what it means, why it matters, how it is detected, the usual reasons with what to check in "
        "the log and whom to ask outside it, the usual actions with countermeasure type and owner role, what to "
        "check first, examples and KPIs, plus the related stage, expectations, failure modes and playbook."
    ),
)
def get_knowledge_hub_page(projectId: str, nodeId: str, c: ContainerDep, process: str | None = None) -> schemas.HubPage:
    return schemas.HubPage(**c.knowledge.hub_page(projectId, nodeId, process))


@router.get(
    "/guidance/{kind}/{entryId:path}",
    operation_id="getGuidance",
    response_model=schemas.Guidance,
    responses={404: {"model": schemas.Problem}},
    description=(
        "The guidance of one expectation area (`layer`), expectation (`constraint` or `expectation`) or failure "
        "mode: the pack's generic tier, the project's own note as `overlay`, and the hub node to open. Pass "
        "`normVersionId` to read the guidance a norm carries with it."
    ),
)
def get_guidance(
    projectId: str,
    kind: Kind,
    entryId: str,
    c: ContainerDep,
    normVersionId: Annotated[str | None, Query(description="read metadata.guidance of this norm version")] = None,
) -> schemas.Guidance:
    return schemas.Guidance(**c.knowledge.guidance(projectId, kind, entryId, norm_version_id=normVersionId))


@router.put(
    "/guidance/{kind}/{entryId:path}",
    operation_id="setGuidanceOverlay",
    response_model=schemas.Guidance,
    responses={404: {"model": schemas.Problem}, 422: {"model": schemas.Problem}},
    description="Your organisation's note on this entry (RK-5). It is added to the generic tier, never instead of it.",
)
def set_guidance_overlay(
    projectId: str, kind: Kind, entryId: str, body: schemas.GuidanceOverlay, c: ContainerDep
) -> schemas.Guidance:
    c.knowledge.set_overlay(projectId, kind, entryId, body.model_dump(exclude_none=True))
    return schemas.Guidance(**c.knowledge.guidance(projectId, kind, entryId))

File: apps/frontend/src/lib/api/knowledge.ts
/** Feature API: knowledge. Generated DTOs remain the wire contract. */

import { queryOptions, useMutation, useQueryClient } from "@tanstack/react-query";
import { http } from "./transport";
import type { components } from "@wise/api-schema";
type S = components["schemas"];
const enc = encodeURIComponent;
const IMMUTABLE = 1000 * 60 * 30;

const knowledgeKeys = {
  hub: (p: string, process: string) => ["projects", p, "knowledge", "hub", process] as const,
  hubPage: (p: string, nodeId: string) => ["projects", p, "knowledge", "hub", "node", nodeId] as const,
  guidance: (p: string, kind: string, id: string) => ["projects", p, "guidance", kind, id] as const,
};

export type GuidanceRef = S["GuidanceRefOut"];

/** One candidate cause of a driver: what to check in the log, or whom to ask outside it. */
export interface UsualReason {
  text: string;
  where?: "log" | "outside" | string;
  check?: string | null;
  reading?: string | null;
}

/** One improvement action of a driver, with the countermeasure type and the role that owns it. */
export interface UsualAction {
  text: string;
  countermeasure?: string | null;
  owner_role?: string | null;
  effect_area?: string | null;
}

export type HubIndex = S["HubIndex"];

export type HubNode = S["HubNode"];

export type HubEdge = S["HubEdge"];

/** The blocks of one guidance tier (`knowledge_hub_panel.md` §1); the contract leaves the object open. */
export interface GuidanceText {
  kind?: string;
  id?: string;
  plain_name?: string | null;
  missed_label?: string | null;
  expectation?: string | null;
  meaning_when_missed?: string | null;
  why_it_matters?: string | null;
  how_detected?: string | null;
  usual_reasons?: UsualReason[];
  usual_actions?: UsualAction[];
  what_to_check_first?: string[];
  examples?: { kind?: string; text?: string; trace?: string | null }[];
  kpis?: string[];
  owner_role?: string | null;
  stakeholders?: string | string[] | null;
  sources?: string[];
  review_status?: string | null;
  version?: string | null;
  note?: string | null;
  author?: string | null;
  updatedAt?: string | null;
  hub_node?: string | null;
  method_name?: string | null;
}

/** The node block of a hub page: the plain name is the title, the method name the subtitle. */
export interface HubNodeFull extends HubNode {
  template?: string | null;
  constraint_id?: string | null;
  constraint_type?: string | null;
  layer?: string | null;
  missed_label?: string | null;
  stage?: string | null;
  order?: number | null;
}

export interface HubRelated {
  stage?: HubNode | null;
  expectations?: HubNode[];
  failure_modes?: HubNode[];
  kpis?: HubNode[];
  reasons?: HubNode[];
  actions?: HubNode[];
  layers?: HubNode[];
  playbook?: { question?: string; id?: string }[];
}

/** Only the open content blocks are refined; named outer fields come from the DTO. */
export type HubPage = Omit<S["HubPage"], "node" | "guidance" | "related" | "overlay"> & {
  node: HubNodeFull; guidance?: GuidanceText | null; related?: HubRelated; overlay?: GuidanceText | null;
};

export type Guidance = Omit<S["Guidance"], "generic" | "overlay"> & { generic?: GuidanceText | null; overlay?: GuidanceText | null };

/** What a hub chip needs to open a page: a node id, or the kind and id the guidance endpoint resolves. */
export interface HubTarget {
  nodeId?: string | null;
  kind?: "layer" | "constraint" | "expectation" | "failure_mode";
  entryId?: string;
  /** What the chip is attached to, for the panel's title before the answer arrives. */
  label?: string;
}

export const hubQuery = (projectId: string, process?: string) =>
  queryOptions({
    queryKey: knowledgeKeys.hub(projectId, process ?? ""),
    queryFn: () => http.get<HubIndex>(`/projects/${enc(projectId)}/knowledge/hub`, { process }),
    staleTime: IMMUTABLE,
    retry: false,
  });

export const hubPageQuery = (projectId: string, nodeId: string, process?: string) =>
  queryOptions({
    queryKey: knowledgeKeys.hubPage(projectId, nodeId),
    queryFn: () => http.get<HubPage>(`/projects/${enc(projectId)}/knowledge/hub/${enc(nodeId)}`, { process }),
    staleTime: IMMUTABLE,
    enabled: !!nodeId,
    retry: false,
  });

export const guidanceQuery = (projectId: string, kind: string, entryId: string, normVersionId?: string) =>
  queryOptions({
    queryKey: [...knowledgeKeys.guidance(projectId, kind, entryId), normVersionId ?? ""] as const,
    queryFn: () => http.get<Guidance>(`/projects/${enc(projectId)}/guidance/${enc(kind)}/${enc(entryId)}`, { normVersionId }),
    staleTime: IMMUTABLE,
    enabled: !!entryId,
    retry: false,
  });

/** *Your organisation's note* (RK-5): added to the pack's text, never instead of it. */
export function useSetOverlay(projectId: string, kind: string, entryId: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (body: { note?: string; author?: string; plain_name?: string }) =>
      http.put<Guidance>(`/projects/${enc(projectId)}/guidance/${enc(kind)}/${enc(entryId)}`, body),
    onSuccess: () => {
      void qc.invalidateQueries({ queryKey: knowledgeKeys.guidance(projectId, kind, entryId) });
      void qc.invalidateQueries({ queryKey: ["projects", projectId, "knowledge"] });
    },
  });
}

/** The hub node an id of the application maps to, for the *What does this mean?* chips. */
export function hubNodeOf(index: HubIndex | undefined, kind: string, id: string): string | undefined {
  if (!index?.nodes?.length) return undefined;
  const wanted = new Set<string>([id, `${kind}:${id}`]);
  for (const node of index.nodes) {
    if (wanted.has(node.id)) return node.id;
    const own = node as HubNodeFull;
    if (kind === "constraint" && own.constraint_id === id) return node.id;
    if (kind === "layer" && own.layer === id) return node.id;
    // node ids are `<kind>:<pack>:<id>`; the last segment is the entry
    const tail = node.id.split(":").slice(-1)[0];
    if (tail === id && node.kind.startsWith(kind === "constraint" ? "expectation" : kind)) return node.id;
  }
  return undefined;
}
